### ISSUE TITLE
Improve UX

### ISSUE THREAD

**Issue by paulmillr** [TablePlus/TablePlus#861]
Thanks for great software.

Coming from Sequel Pro, a popular open-source MySQL client.

Immediately noticed TablePlus was designed by engineer-like mindset. I think it would be great to clean-up the UI a bit and make it more intuitive. Current UI wastes A LOT of user time. For example, it is impossible to get back from "SQL" tab. It's not even a tab. It should be a tab like Data and Structure.

See 1) — clean shot of Sequel Pro, and 2) — some recommendations on UX improvement.
 
<img width="1248" alt="screen shot 2018-12-25 at 2 09 55 am" src="https://user-images.githubusercontent.com/574696/50408182-fa08c380-07ee-11e9-8fb0-2bf504fa0313.png">
<img width="1332" alt="marked" src="https://user-images.githubusercontent.com/574696/50408183-ff660e00-07ee-11e9-92d3-54f714d22eec.png">

**Comment by paulmillr** [comment]
My point is: most complex software can be made more structured and provide great UX. Even if you have 5x more features vs competitor. The suggestions here are just whatever came to mind first. The app does not feel awesome and does not feel native.

For a real work, a real UX designer should be hired. He can work with you directly and you'll discuss improvement path. I think it's very reasonable — at least if you expect users pay $50 for the software.

**Comment by alvint** [comment]
I imagine this is all a matter of opinion. All UXs are a series of judgement calls and compromises; what works for one person may not for another. All I can say is my opinion: I've been a software engineer for over 20 years and I've used literally dozens of DB clients. I switched to TP precisely because its UX better matches my workflow. It's not perfect but it's a solid start.

**Comment by paulmillr** [comment]
Totally agree re: a solid start. Even in its current state it's much better than navicat and all the crap out there.

Regarding "a matter of opinion" — not really. UI — yes, UX — not really. There are some factual UX rules that allow to construct better software. See books like "Don't Make Me Think" & "Design of Everyday Things". 

What i'm talking about is not "replacing icons" or moving them around, it's studying the common workflows and making experience pleasurable and more effective for most cases.

**Comment by huyphams** [comment]
Thanks @alvint and @paulmillr I agreed with you but in my case, I'm not only based on the common rule UX on those books. TablePlus's design is based on community and user feedbacks. That's why we're here. I will rethink/fix it if there are a lot of people in trouble with the feature. But everything is fine and easy understand with all people.
TablePlus don't want to be another one, it has its own design. I'm not using Sequel Sequel before starting TablePlus so I don't see any problem (Probably, because I'm using PostgreSQL all the time).
I have downloaded Sequel Pro recently and I am a little bit confused about it because it always shows the toolbar (like the old browser and Navicat IMO).

**Comment by devtimi** [comment]
I agree with most of the changes @paulmillr pointed out. I think these would greatly simplify the UI and improve the UX by making the software easier to understand. (I think the row details view can stay, but there are probably better ways of conveying the data types without taking up so much space.)

Part of the issue with the current UX is that unrelated items are grouped together, and placed in very weird spots. It's not an easily discoverable UI.

Two things I would suggest that would bring the UI learning curve much lower:
- Raw SQL entry is very important. "SQL Query" should be _much_ easier to get to and find. There's not even a menu item, so we can't use System Preferences to assign it a familiar shortcut. We're stuck with whatever is hard coded. I recall spending quite some time trying to figure out where this was when I first opened TablePlus.
- Database connections should not all be one window with _yet another_ tab thing on the left. There are too many tabs, and too many _types of tabs_. I might suggest that you have a solid use paradigm for example, Connection = window, view/table = tab. As flat as Jony Ive wants the world to be, there's still use to multiple windows. But that's not to say you have to take that paradigm. Sequel Pro for example makes a connection a tab, and everything is handled _within_ that tab. (The downside being that viewing two tables at once needs two connections).

The concern with "always showing the structure and the data button for all item types, including function and procedure. it made no sense." is actually backwards from discoverability. As Apple suggests in the HIG for tabbed views, if there's nothing to do on a tab (like say, it's not available for a procedure) it should still be easy to find and explain on that view why there's nothing to do there.

> If tabs are available in some cases but not in others, your app’s interface becomes unstable and unpredictable. Ensure that all tabs are always enabled, and explain why a tab’s content is unavailable. For example, if there are no songs on an iOS device, the My Music tab in the Music app explains how to download songs.
https://developer.apple.com/design/human-interface-guidelines/ios/bars/tab-bars/

I love improving UI/UX, so I'd be happy to suggest more; and as above I can justify each of my suggestions with backed time-tested paradigms. But I know it's hard to receive criticism on an interface you've designed and like, so I don't want to overdo it.

**Comment by chenjau** [comment]
> But everything is fine and easy understand with all people.

Are you sure? sorry but I have to say that the UX just crap.

**Comment by huyphams** [comment]
Happy new year @chenjau

**Comment by chenjau** [comment]
@huyphams Happy new year! 
Hope you guys realise that not ALL people happy with current UX design and find room for improvement.

**Comment by huyphams** [comment]
I made TablePlus a year ago and I have learned something.
1. Even you made a free tool, people are still yelling at you and sometimes it's really bad (and believe me, you have no idea how bad can it be).
2. You can't make a tool like a silver bullet that can solve everything and fix with everybody.
3. If you have enough users, everything you do wrong can results with a lot of feedbacks and emails every day. I've got a lot of users and I have/know that data, but you don't. And with me, TablePlus isn't focusing on table structure, edit structure... so there is no reason I would mess up the toolbar with it. Sublime Text, VSCode don't even have a toolbar.
4. I appreciate all the feedback but I must have a strong opinion because many people have supported because they believe me. If I make a change based on a small group of people, all the rest will email and yelling at me.

iOS, macOS is great, but Windows and Android fan still saying it's crap, again I can't make a silver bullet that all people can love. Forgive me because English is not my native language :D

I appreciate all the feedback and will keep thinking about it. 

Happy new year everybody 🎉

**Comment by devtimi** [comment]
I hope you reconsider "TablePlus isn't focusing on table structure, edit structure... so there is no reason I would mess up the toolbar with it.". I wouldn't consider database structure "clutter" at all, especially when managing databases.

To help explain what I mean - some structures are designed by the less-inclined, and don't all fit onscreen. In these instances I find that viewing the structure helpful in finding the column I actually need. TablePlus doesn't need to be Navicat in scale of database structuring, but a database tool that dismisses structure is dismissible itself.

I wish you a happy new year @huyphams, and apologize for the way some users are. Your English is certainly better than my... well any other language.

**Comment by vesper8** [comment]
Hi @huyphams, I clicked on this thread because earlier today I actually bought your software and even had a nice conversation with you on your website chat. And as a result of this I have now dug a little deeper into using TablePlus and letting it slowly become my main SQL Client on OSX. So count me as one of the, hopefully many, who not only appreciate your freeware software but liked it enough that they are now supporting you with their money. 

I applaud you for the amazing work you've done so far. I see you focus a lot on adding new features, new drivers. But I also see you pushing many of these features quickly and then having half-finished and sometimes hardly-working features. 

I'm thinking of the MongoDB implementation which I just tried again this very second in build 207. I'm sorry to say I found many unintuitive things with it just two minutes of trying to use it. I guess I should create another issue for it or add it to my thread of MongoDB enhancement requests. At this point they are not so much enhancements but I would advise you to pull it out completely until it is more functional.. it was pushed out way too soon, it doesn't even work at all today, can't even list tables/collections after successfully connecting to a database.

Ok sorry for dragging on, it's just that now that I'm using it more I'm also finding a lot bugs and a lot of things that are bad UX, and I came here hoping to see you discuss out whatever was being written rather than act defensively about it. The UX needs a lot of work, there are a lot of usability bugs at the moment, some things are great but lots of things are still lacking and you have to understand that most developers that would use a single software with multiple drivers (redis, mysql, mongo) already have favorites for each of those.. You're trying to be a jack of all trades is making it very hard for you to reproduce the feature sets of existing apps that specialize in only a single driver.

The crazy thing is, I like TablePlus and bought it for one reason only, the ability to 'open anything' with CMD-P and then I can type in a database or a table. Being someone who works with 60+ databases and 1000s of tables, this feature is priceless to me and is worth all in itself the price of entry. But for now I continue to use Studio 3T, Sequel Pro and Medis in parallel because this app still needs a lot of polish, a lot of bug fixing instead of focussing on new features, and a lot of user experience improvements. And also you shouldn't talk badly of Sequel Pro considering it's made for free by a community of developers and it's more mature and polished than TP IMHO. It may lack features but we don't all need so many features, we just want to be able to do the simple things we do on a daily basis as easily and pleasantly as possible.

And up there you say "TablePlus's design is based on community and user feedbacks." as if to say "your feedback is not relevant because many users of the community haven't expressed it yet".. are you hearing yourself? We ARE the community right here in this thread, taking the time to post issues trying to have this app improve because we care about it and like it enough to bother posting about it. Not everyone file bug/feedback issues obviously.. those that do and approach you talking in a civil manner, trying to make a point (valid points above that I mostly agree with by the way), we are the ones helping you make your app better so we can be happier and you can be more proud of your accomplishments and hopefully make money and win more recognition.

And as a paying customer, I do wish you would be less defensive and more open to things such as hiring some help instead of trying to do everything yourself. I'm a dev too and we all know how it is, we think we know everything and think we know best but anyone that's been around long enough knows that's more ego than fact.. we are bigger winners when we work together and aren't afraid to ask for help or admit that we could use it.

Happy NYE a few months late ;) Keep up the good work. No offense meant here, I like you and TP and look forward to what it will become. Off to a great start indeed!

**Comment by huyphams** [comment]
Thanks, @vesper8 for the honest feedback. I appreciate your feedback 👍 

Regarding the Mongo: yes, it's a beta release and not fully support (I'm working on the JSON Viewer) and Mongo Shell feature which will compatible with Mongo Shell Query (such as db.find())
`it doesn't even work at all today, can't even list tables/collections after successfully connecting to a database.` => I think there is something wrong with the URL connection string.

UI, UX: yes I'm redesigning, for example, creating new database now has encoding and collation, it will support tables and columns as well. It will compatible with Sequel Pro.

<img width="615" alt="Screen Shot 2019-03-27 at 8 19 38 AM" src="https://user-images.githubusercontent.com/3654622/55043480-1411a680-5069-11e9-983d-96e9b68a3023.png">

And I'm sure that I will take seriously the UX enhancement over new features.

P.S. We're a team now, not only me anymore 😄

**Comment by gennaios** [comment]
Agree with some of the points. New query is difficult to get to as well as switching to table structure. I don't necessarily mind that data/structure etc are in the bottom toolbar – seems like a good place –, though some menu item for table operations and views could be helpful, as well as adding them to Open Anything.

Open Anything
- could include all connections; open anything should include anything and I think it's more intuitive as an addition than shift-cmd-K. one could have multiple tabs, connections, etc., and one is used to cmd-P to switch.
- transparency when the dialog is open is strange. Other apps that implement similar such as OmniFocus, Ulysses, Intellij IDEs, etc., they don't do this. there's a slight shadow with the dialog, that's it. I think it'd look much nicer.
- connections, if they show up, open anything could show all options for all open connections. such as connection1 -> table, …, etc. switching to anything open in the app would be great. even unopen connections that were previously opened, it should know what options are available.

Just started using TablePlus and it's great. those are somethings I noticed so far.

**Comment by mrmurphy** [comment]
Just wanted to pipe in and say thanks to you @huyphams. I'm overall very happy with Table Plus, and though there are some friction points, the user interface has been in general very nice and discoverable for me.
